### PR TITLE
chore: check diamond contract storage changes on pr

### LINKED ISSUE #636: check that diamond storage is untouched after each pr
best would be to add a gh action to make sure that we haven't messed with the diamond storage

### PR DESCRIPTION
**Pull request by gitcoindev**
Resolves: #636

The solution described at https://github.com/ubiquity/ubiquity-dollar/issues/636#issuecomment-1715174347

### CODE REVIEW

**Comment by molecula451** [comment]
Rename the GitHub Action to Check For Diamond Storage Changes

**Comment by gitcoindev** [comment]
> Rename the GitHub Action to Check For Diamond Storage Changes

Will be done.

**Comment by gitcoindev** [comment]
> > Rename the GitHub Action to Check For Diamond Storage Changes
> 
> Will be done.

done!

**Comment by gitcoindev** [comment]
I renamed both GitHub action name and step for consistency.

**Comment by gitcoindev** [comment]
Hi @rndquu @molecula451 thank you for reviewing and testing the core part. After this one is tested and approved, we could as well move my [fork](https://github.com/gitcoindev/foundry-storage-check) under [ubiquity](https://github.com/ubiquity) or some other location if you would like to keep more control of this repository in the future.

**Comment by rndquu** [comment]
@gitcoindev Could you elaborate on why [using a mapping](https://github.com/Rubilmax/foundry-storage-check/compare/main...gitcoindev:foundry-storage-check:main#diff-a9aa7d8f1bead33c53fb22bf0cdc4dfed1ac646ff4fbd59ac789b9b52770d3c0R218) inside a storage struct is considered not upgrade safe? As far as I understand using mappings in a storage struct is fine and upgrade safe.

**Comment by gitcoindev** [comment]
> @gitcoindev Could you elaborate on why [using a mapping](https://github.com/Rubilmax/foundry-storage-check/compare/main...gitcoindev:foundry-storage-check:main#diff-a9aa7d8f1bead33c53fb22bf0cdc4dfed1ac646ff4fbd59ac789b9b52770d3c0R218) inside a storage struct is considered not upgrade safe? As far as I understand using mappings in a storage struct is fine and upgrade safe.

Hi @rndquu this is a great question. Yes in my view it is considered upgrade safe to use mappings. This test case is about inserting a variable in the middle of the storage struct, similar to primitive variables.

I added this test case to make sure that the storage layout can be also generated for a storage with mappings. When I tested simple structs, the tool worked correctly. When I ran it against structs with mappings it threw an error.

For example for LibBondingCurve.sol it gave:

[check_storage_layout (src/dollar/libraries/LibBondingCurve.sol:LibBondingCurve)](https://github.com/gitcoindev/ubiquity-dollar/actions/runs/6208699157/job/16855126839#step:4:41)
TypeError: Cannot read properties of undefined 

the `TypeError` was caused by inability of parsing this line https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/src/dollar/libraries/LibBondingCurve.sol#L42C35-L42C43

I fixed this, and added a test case to prove that mappings can be used.

If the mapping is inserted not in the middle, but at the end of the storage struct, the tool should not throw an error.

I hope I described this clearly, if something is still not clear please let me know.

**Comment by rndquu** [comment]
@gitcoindev 

Works fine for the following cases:
- success: new variable is added to the end of a storage struct
- fail: new variable introduced in the beginning or in the middle of a storage struct
- fail: type of state variable changed
- fail: existing state variable removed from the middle of a storage struct
- fail: LibAppStorage should follow the same diamond layout safe rules

The following cases didn't work as expected (or there are questions regarding them):
- fail: order of state variables changed 
- fail: existing state variable removed from the end of a storage struct
- fail: inner struct is added to a storage struct. Do not put structs directly in another struct unless you don't plan on ever adding more state variables to the inner structs. You won't be able to add new state variables to inner structs in upgrades without overwriting the storage slot of variables declared after the struct. The solution is to add new state variables to the structs that are stored in mappings, rather than putting structs directly in structs, as the storage slot for variables in mappings is calculated differently and is not continuous in storage.
- fail: do not add new state variables to structs that are used in arrays. Example:
```
struct AppStorage {
  uint totalSupply;
  DoNotUpgrade[] public myArray;
}

struct DoNotUpgrade {
  ...
}
```

Questions:
1. Why do we need to [inspect](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L26) the `DiamondInit` contract? It is not part of any facet neither it is UUPS upgradeable.
2. This [command](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L28) doesn't include the [LibAccessControl](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L28) library in the result output. Why is it happening?
3. Why do we [include](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L27) the facets in the result output? Isn't it sufficient to include only the libraries?
4. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/95e6b6bad198acc501bb538a582e47385aba4c82) where state variables order was updated. The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300083738/job/17102154494) didn't throw any errors, only warnings. Can we throw an error in this case? It is really easy to miss the warning.
5. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/718c4c43082b8c3266b4f30681fbfe90032f4d04) where state variable was removed from the end of the storage struct. The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300456922/job/17103385221) didn't throw any errors or warnings. Can we throw an error in this case?
6. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/2aa5542ecdc6cb4c1c7f75e8a13299ecd809ae90) where an inner struct is declared inside a storage struct. Generally we should avoid adding inner structs to storage structs because this way inner structs become not upgrade safe (i.e. we can't add new state variables to inner structs). The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300584619/job/17104492043) didn't throw any errors or warnings. On the other hand we can still use inner structs (actually we already [use](https://github.com/rndquu/ubiquity-dollar/blob/1f8fbf80b714f4c7ebbf3911e5e451530b4f146a/packages/contracts/src/dollar/libraries/LibChef.sol#L53) them in some of the libs) if we 100% sure that we won't add new state variables to inner structs so I'm in doubt if we need to throw an error or simply to keep an eye on such things. What do you think?
7. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/5cd474d6e1d22fdd9fabfd0f8e45adb002a5ecd6) where a dynamic array of structs is added. The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300919935/job/17104827148) didn't throw any errors or warnings. This [article](https://www.certik.com/resources/blog/7laIe0oZGK6IoYDwn0g2Jp-diamond-proxy-contracts-best-practices) states that we should not add new state variables to structs that are used in arrays. I don't really understand this example because from my point of view this code inside a struct `DoNotUpgrade[] public myArray;` is upgrade safe. What do you think?

P.S. Note to self: run storage check workflows only when `*.sol` files are updated

**Comment by gitcoindev** [comment]
Hi @rndquu thank you for testing and the detailed feedback. It does not look so bad overall. I will reply one by one.

> Questions:
> 
> 1. Why do we need to [inspect](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L26) the `DiamondInit` contract? It is not part of any facet neither it is UUPS upgradeable.

After some more thought I agree and solved this in https://github.com/ubiquity/ubiquity-dollar/pull/788/commits/07fd6ed2eae44951242858228041fb7863f8ab60

> 2. This [command](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L28) doesn't include the [LibAccessControl](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L28) library in the result output. Why is it happening?

Solved in https://github.com/ubiquity/ubiquity-dollar/pull/788/commits/05e20b500aba5a425d0f587d49ff1e41854fd5ba

> 3. Why do we [include](https://github.com/ubiquity/ubiquity-dollar/blob/9d70040f76b5dbcba3ff7d05b94feffc0cc4640a/.github/workflows/diamond-storage-check.yml#L27) the facets in the result output? Isn't it sufficient to include only the libraries?

After some more thought I agree and solved this in https://github.com/ubiquity/ubiquity-dollar/pull/788/commits/07fd6ed2eae44951242858228041fb7863f8ab60

> 4. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/95e6b6bad198acc501bb538a582e47385aba4c82) where state variables order was updated. The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300083738/job/17102154494) didn't throw any errors, only warnings. Can we throw an error in this case? It is really easy to miss the warning.

I implemented a new optional feature for failOnLabelDiff, see

https://github.com/Rubilmax/foundry-storage-check/commit/95e9b089dc6bcb1c98cbbf56e9bc8bb5e81170fc
https://github.com/Rubilmax/foundry-storage-check/commit/e2e33e2073c0bf9a5b35124bc117382442b54c2d

released as version 3.6.3 and enabled it at

https://github.com/ubiquity/ubiquity-dollar/pull/788/commits/ad609eb99883ea2791138361ad57ea619eb85ca1

> 5. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/718c4c43082b8c3266b4f30681fbfe90032f4d04) where state variable was removed from the end of the storage struct. The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300456922/job/17103385221) didn't throw any errors or warnings. Can we throw an error in this case?

I enabled failOnRemoval at

https://github.com/ubiquity/ubiquity-dollar/pull/788/commits/34baaece98ba41a063eb6047c2427f99868bb7d1

> 6. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/2aa5542ecdc6cb4c1c7f75e8a13299ecd809ae90) where an inner struct is declared inside a storage struct. Generally we should avoid adding inner structs to storage structs because this way inner structs become not upgrade safe (i.e. we can't add new state variables to inner structs). The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300584619/job/17104492043) didn't throw any errors or warnings. On the other hand we can still use inner structs (actually we already [use](https://github.com/rndquu/ubiquity-dollar/blob/1f8fbf80b714f4c7ebbf3911e5e451530b4f146a/packages/contracts/src/dollar/libraries/LibChef.sol#L53) them in some of the libs) if we 100% sure that we won't add new state variables to inner structs so I'm in doubt if we need to throw an error or simply to keep an eye on such things. What do you think?

I agree, I would keep an eye on such things as well as inner structs are already used.

> 7. Check this [commit](https://github.com/rndquu/ubiquity-dollar/pull/30/commits/5cd474d6e1d22fdd9fabfd0f8e45adb002a5ecd6) where a dynamic array of structs is added. The [corresponding CI run](https://github.com/rndquu/ubiquity-dollar/actions/runs/6300919935/job/17104827148) didn't throw any errors or warnings. This [article](https://www.certik.com/resources/blog/7laIe0oZGK6IoYDwn0g2Jp-diamond-proxy-contracts-best-practices) states that we should not add new state variables to structs that are used in arrays. I don't really understand this example because from my point of view this code inside a struct `DoNotUpgrade[] public myArray;` is upgrade safe. What do you think?
> 

To be honest I also share your opinion, the code looks upgrade safe.

> P.S. Note to self: run storage check workflows only when `*.sol` files are updated

I already solved this in commit:

https://github.com/ubiquity/ubiquity-dollar/pull/788/commits/93f6a57ba9cc04f3e07f49bfb60085cfbb41dc3f

**Comment by gitcoindev** [comment]
@rndquu I answered all the questions and implemented new functionalities and fixes. I would be grateful if you had a look again at the PR.

File: .github/workflows/diamond-storage-check.yml
Comment (rndquu) [inline .github/workflows/diamond-storage-check.yml]: Pls [refactor](https://github.com/ubiquity/ubiquity-dollar/pull/795/files#diff-44b0878aa3a2c194318bbbbde7166259a120f3ad0a3497153359dd03db11f32fR27) to `GITHUB_OUTPUT`

**Review by rndquu — CHANGES_REQUESTED** [review]
@gitcoindev 

All of the [comments](https://github.com/ubiquity/ubiquity-dollar/pull/788#issuecomment-1735096038) are addressed, everything works fine, thank you

Could you move the https://github.com/gitcoindev/foundry-storage-check action to the [ubiquity](https://github.com/ubiquity) organization? Not sure if you have enough permissions though.

**Comment by rndquu** [comment]
@gitcoindev Could you also elaborate on [this line](https://github.com/gitcoindev/foundry-storage-check/blob/e2e33e2073c0bf9a5b35124bc117382442b54c2d/src/input.ts#L43). Why `child.type === "ContractDefinition"` is considered to be a library?

File: .github/workflows/diamond-storage-check.yml
Comment (gitcoindev) [reply .github/workflows/diamond-storage-check.yml]: done!

**Comment by gitcoindev** [comment]
> @gitcoindev Could you also elaborate on [this line](https://github.com/gitcoindev/foundry-storage-check/blob/e2e33e2073c0bf9a5b35124bc117382442b54c2d/src/input.ts#L43). Why `child.type === "ContractDefinition"` is considered to be a library?

Yes, sure. This is an internal representation in Solidity AST (Abstract Syntax Tree). The contract vs library is recognized later in contractKind field, which can be set to : "contract" | "interface" | "library" . 

I also fixed the GitHub action warning with the same solution as for the core contracts.
I am quite happy that the solution works!

Unfortunately I do not have permissions to move https://github.com/gitcoindev/foundry-storage-check to Ubiquity GitHub, but perhaps @pavlovcik can do this.

**Comment by 0x4007** [comment]
https://github.com/ubiquity/foundry-storage-check open a pr

**Comment by gitcoindev** [comment]
> https://github.com/ubiquity/foundry-storage-check open a pr

GitHub does not seem to allow me. But I have a simple solution, please just "fork my fork" -) by clicking on this link:

https://github.com/gitcoindev/foundry-storage-check/fork

**Comment by molecula451** [comment]
It's be to fork the repo to be part of the org than get into dollar repo. It could be use as a submodule

**Comment by gitcoindev** [comment]
Hi @molecula451 in other words do you suggest to fork or not to fork?

**Comment by molecula451** [comment]
Yes to fork instead of a PR

**Comment by gitcoindev** [comment]
Ok, thank you. @pavlovcik you can then simply remove the empty project https://github.com/ubiquity/foundry-storage-check , click on the link https://github.com/gitcoindev/foundry-storage-check/fork when you are logged in to Ubiquity GitHub. After selecting Ubiquity as the destination GitHub project, the repository should be forked in a few seconds.

**Comment by 0x4007** [comment]
Then when you delete your source ours delete as well. Just push the code to the repo so we don't have issues in the future

**Comment by gitcoindev** [comment]
@pavlovcik ok I see. I tried but I cannot push neither to 'main' branch or to a 'feature' branch there. I think GitHub prevents me due to lack of permissions.

```
09:12:40 (main) foundry-storage-check$ git push --set-upstream origin main
ERROR: Permission to ubiquity/foundry-storage-check.git denied to gitcoindev.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

09:12:58 (main) foundry-storage-check$ git co -b feature/ubiquity
Switched to a new branch 'feature/ubiquity'

09:13:42 (feature/ubiquity) foundry-storage-check$ git push --set-upstream origin feature/ubiquity
ERROR: Permission to ubiquity/foundry-storage-check.git denied to gitcoindev.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

It seems that only you can do this.

You would need to execute the following:

```
$ git clone git@github.com:ubiquity/foundry-storage-check.git
$ cd foundry-storage-check/
$ git remote add gitcoindev git@github.com:gitcoindev/foundry-storage-check.git
$ git fetch gitcoindev
$ git checkout gitcoindev/main
$ git checkout -b main
$ git push --set-upstream origin main
```

**Comment by 0x4007** [comment]
Open a pr

**Comment by rndquu** [comment]
I think we should fork the original repo https://github.com/Rubilmax/foundry-storage-check to the ubiquity org and then @gitcoindev could open a PR with the changes. This way we could sync new updates from the [original](https://github.com/Rubilmax/foundry-storage-check) repo.

I've pushed the changes to https://github.com/ubiquity/foundry-storage-check but the repo is not forked so we can not track updates. I don't have enough rights to delete the https://github.com/ubiquity/foundry-storage-check repo, @pavlovcik could you delete it? I will then fork the original https://github.com/Rubilmax/foundry-storage-check repo to the ubiquity org.

P.S. If original repo is deleted then forks [remain untouched](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility#deleting-a-public-repository)

**Comment by rndquu** [comment]
@gitcoindev Any additions [here](https://github.com/ubiquity/ubiquity-dollar/wiki/23.-Smart-Contracts-Architecture#upgradeability)?

**Comment by gitcoindev** [comment]
> @gitcoindev Any additions [here](https://github.com/ubiquity/ubiquity-dollar/wiki/23.-Smart-Contracts-Architecture#upgradeability)?

Looks pretty good, even has inner structs explanation. Perhaps at some point a list of best practices with examples on how to upgrade could be created out it, but this is a subject for a new wiki page or a lenghty medium article :)

**Comment by gitcoindev** [comment]
@pavlovcik since the pull request was tested and https://github.com/ubiquity/foundry-storage-check contains a copy of my fork, can the issue be approved and closed? If yes, I will have a look at devpool and pick a new task. 

I can also do what @rndquu suggested if you decide to delete https://github.com/ubiquity/foundry-storage-check , fork from original then I will open a pr from my fork.

**Comment by 0x4007** [comment]
pr to here https://github.com/ubiquity/foundry-storage-check

**Comment by gitcoindev** [comment]
> pr to here https://github.com/ubiquity/foundry-storage-check

https://github.com/ubiquity/foundry-storage-check/pull/1

File: .github/workflows/diamond-storage-check.yml
Comment (rndquu) [inline .github/workflows/diamond-storage-check.yml]:
```suggestion
        uses: ubiquity/foundry-storage-check@main
```

**Review by rndquu — CHANGES_REQUESTED** [review]

**Review by rndquu — APPROVED** [review]

File: .github/workflows/diamond-storage-check.yml
Comment (gitcoindev) [reply .github/workflows/diamond-storage-check.yml]: just saw the commit from you, thank you!

**Comment by gitcoindev** [comment]
Thank you for approval @rndquu ! I would be grateful if you @molecula451 also had a look and check if can be approved from your side.

**Review by molecula451 — APPROVED** [review]
This PR meets all the specifications required at #636.
The hunter went thru all the asked question and applied changes accordingly.
This PR is of highest excellence, because it meets simplicity and complexity at the same time.

There is no need for other changes that this PR requires other that further changes that might come along in the future, with diamonds, storage, etc. Therefore i am approving the PR.

My QA:

Workflow triggered

https://github.com/molecula451/ubiquity-dollar/actions/runs/6335696942

this PR can be safely merge

**Pull request merged**
